src/library/scala/collection/Iterator.scala

```diff
@@ -1189,6 +1223,80 @@ object Iterator extends IterableFactory[Iterator] {
       } else Iterator.empty.next()
     }
   }
+
+  private final class TakeRightIterator[A](private[this] var underlying: Iterator[A], buflen: Int) extends AbstractIterator[A] {
+    private[this] var len: Int = -1
+    private[this] var pos: Int = 0
+    private[this] var buf: Array[AnyRef] = _
+    def init(): Unit = if(buf eq null) {
+      buf = new Array[AnyRef](buflen)
+      len = 0
+      while(underlying.hasNext) {
+        buf(pos) = underlying.next().asInstanceOf[AnyRef]
+        pos = (pos + 1) % buflen
+        len += 1
+      }
+      underlying = null
+      if(len > buflen) len = buflen
+      pos = (pos - len) % buflen
+      if(pos < 0) pos += buflen
+    }
+    override def knownSize = len
+    def hasNext: Boolean = {
+      init()
+      len > 0
+    }
+    def next(): A = {
+      init()
+      if(len == 0) Iterator.empty.next()
+      else {
+        val x = buf(pos).asInstanceOf[A]
+        pos = (pos + 1) % buflen
+        len -= 1
+        x
+      }
+    }
+    override def drop(n: Int): Iterator[A] = {
```

Comment:
Why just `drop`?  Now that everything is in a ring buffer, `take`, `takeRight`, `slice`, etc. all have efficient implementations.  Is there something special about `drop` that makes it more likely to be used than the others?

Reply:
I think it's because `it.takeRight(n).drop(k)` is equivalent to `it.takeRight(n - k)`, and this iterator is already an implementation for `takeRight`. On the other hand,

- `it.takeRight(n).take(k)` is only equivalent to `it.takeRight(x)` if `k >= n` (in which case `x = n`)
- `it.takeRight(n).slice(a, b)` is only equivalent to `it.takeRight(x)` if `b > n` (in which case `x = b - a`)

However, `it.takeRight(n).takeRight(k)` *is* equivalent to `it.takeRight(x)`, where `x = n min k`, so `takeRight` should probably be overridden as well.

(Disclaimer: there is probably at least one off-by-one error in the above; use with caution)

Reply:
`drop` was easy to implement. With too many other PRs to look after, I'll leave further optimizations to others.